File: 2023/day7/day7.clj
(require '[clojure.string :as str])

(def input-raw (slurp (-> *file* (str/split #"/") butlast (#(str/join "/" %)) (str "/input.txt"))))

#_(def input-raw "
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483")

(def input
  (-> input-raw
      str/trim
      (str/split-lines)))

(defn includes? [list val] (contains? (set list) val))

(defn parse-hand [line]
  (let [[cards bid] (str/split line #" ")]
    {:cards cards :bid (parse-long bid)}))

(defn get-type [cards]
  (let [freq (frequencies cards)
        values (vals freq)
        key-count (count (keys freq))]
    (cond
      ;; 5 of a kind
      (includes? values 5) 7
      ;; 4 of a kind
      (includes? values 4) 6
      ;; Full house
      (and (includes? values 3) (= key-count 2)) 5
      ;; 3 of a kind
      (and (includes? values 3) (= key-count 3)) 4
      ;; 2 pair
      (and (includes? values 2) (= key-count 3)) 3
      ;; 1 pair
      (and (includes? values 2) (= key-count 4)) 2
      :else 1)))

(defn by-rank [hand1 hand2]
  (let [type-diff (compare (get-type (:cards hand1)) (get-type (:cards hand2)))]
    (if (zero? type-diff)
      (let [strength ["AKQJT98765432"]
            strength1 (->> (:cards hand1) vec (map str) (map (partial str/index-of strength)))
            strength2 (->> (:cards hand2) vec (map str) (map (partial str/index-of strength)))
            strength-diff (map #(apply - %1) (partition 2 (interleave strength2 strength1)))]
        (first (drop-while zero? strength-diff)))
      type-diff)))

(defn get-best-hand [hand]
  (let [freq (frequencies hand)
        j (get freq \J 0)]
    (if (= j 5)
      "AAAAA"
      (let [freq-pairs (->> freq (into [] cat) (partition 2) (filter #(not= (first %) \J)) (sort-by second) reverse)
            add-j (concat [[(ffirst freq-pairs) (+ j (second (first freq-pairs)))]] (rest freq-pairs))
            new-hand (apply str (mapcat (fn [[s n]] (repeat n s)) add-j))]
        new-hand))))

(defn by-rank-wild-j [hand1 hand2]
  (let [type-diff (compare (get-type (get-best-hand (:cards hand1))) (get-type (get-best-hand (:cards hand2))))]
    (if (zero? type-diff)
      (let [strength ["AKQT98765432J"]
            strength1 (->> (:cards hand1) vec (map str) (map (partial str/index-of strength)))
            strength2 (->> (:cards hand2) vec (map str) (map (partial str/index-of strength)))
            strength-diff (map #(apply - %1) (partition 2 (interleave strength2 strength1)))]
        (first (drop-while zero? strength-diff)))
      type-diff)))

(defn part1 [input]
  (->> input
       (map parse-hand)
       (sort by-rank)
       (map-indexed #(* (inc %1) (:bid %2)))
       (reduce +)))

(defn part2 [input]
  (->> input
       (map parse-hand)
       (sort by-rank-wild-j)
       (map-indexed #(* (inc %1) (:bid %2)))
       (reduce +)))

(println "part1:" (part1 input))
(println "part2:" (part2 input))

File: 2023/day7/day7.ts
interface Hand {
  cards: string;
  bid: number;
}

function frequencies(arr: string[]) {
  return arr.reduce(function (value, value2) {
    return value[value2] ? ++value[value2] : (value[value2] = 1), value;
  }, {});
}

function getType(cards: string): number {
  let freq = frequencies(cards.split(''));
  if (Object.values(freq).includes(5)) {
    // 5 of a kind
    return 7;
  } else if (Object.values(freq).includes(4)) {
    // 4 of a kind
    return 6;
  } else if (Object.values(freq).includes(3) && Object.keys(freq).length === 2) {
    // Full house
    return 5;
  } else if (Object.values(freq).includes(3) && Object.keys(freq).length === 3) {
    // 3 of a kind
    return 4;
  } else if (Object.values(freq).includes(2) && Object.keys(freq).length === 3) {
    // 2 pair
    return 3;
  } else if (Object.values(freq).includes(2) && Object.keys(freq).length === 4) {
    // 1 pair
    return 2;
  } else {
    // High card
    return 1;
  }
}

function byRank(hand1: Hand, hand2: Hand): number {
  let typeDiff = getType(hand1.cards) - getType(hand2.cards);
  if (typeDiff !== 0) return typeDiff;

  let strength = ['A', 'K', 'Q', 'J', 'T', '9', '8', '7', '6', '5', '4', '3', '2'];
  for (let i = 0; i < hand1.cards.length; i++) {
    let strengthDiff = strength.indexOf(hand2.cards[i]) - strength.indexOf(hand1.cards[i]);
    if (strengthDiff !== 0) return strengthDiff;
  }

  return typeDiff;
}

function getBestHand(cards: string) {
  let freq = frequencies(cards.split(''));
  if (freq['J'] === undefined) {
    return cards;
  } else {
    let j = freq['J'];
    delete freq['J'];
    let entries = Object.entries(freq) as [string, number][];
    if (entries.length === 0) {
      return 'AAAAA';
    }
    entries.sort((a, b) => b[1] - a[1]);
    entries[0][1] += j;
    let newHand = entries.map(e => e[0].repeat(e[1])).join('');
    return newHand;
  }
}

function byRankWildJ(hand1: Hand, hand2: Hand): number {
  let typeDiff = getType(getBestHand(hand1.cards)) - getType(getBestHand(hand2.cards));
  if (typeDiff !== 0) return typeDiff;

  let strength = ['A', 'K', 'Q', 'T', '9', '8', '7', '6', '5', '4', '3', '2', 'J'];
  for (let i = 0; i < hand1.cards.length; i++) {
    let strengthDiff = strength.indexOf(hand2.cards[i]) - strength.indexOf(hand1.cards[i]);
    if (strengthDiff !== 0) return strengthDiff;
  }

  return typeDiff;
}

function part1(input: string) {
  const lines = input.trim().split('\n');
  let rounds = [] as Hand[];
  for (let line of lines) {
    let [cards, bidStr] = line.split(' ');
    let bid = Number(bidStr);
    rounds.push({ cards, bid });
  }
  rounds.sort(byRank);
  return rounds.map((hand, rank) => hand.bid * (rank + 1)).reduce((a, b) => a + b);
}

function part2(input: string) {
  const lines = input.trim().split('\n');
  let rounds = [] as Hand[];
  for (let line of lines) {
    let [cards, bidStr] = line.split(' ');
    let bid = Number(bidStr);
    rounds.push({ cards, bid });
  }
  rounds.sort(byRankWildJ);
  return rounds.map((hand, rank) => hand.bid * (rank + 1)).reduce((a, b) => a + b);
}

let input = await Bun.file(`${import.meta.dir}/input.txt`).text();

let testInput = `
32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483
`;
// input = testInput;

console.log('part1:', part1(input));
console.log('part2:', part2(input));
